src/libraries/System.IO.Compression.Zstandard/src/System/IO/Compression/ZstandardStream.cs

```diff
@@ -0,0 +1,240 @@
+        /// <param name="value">The desired length of the current stream in bytes.</param>
+        /// <exception cref="NotSupportedException">This property is not supported on this stream.</exception>
+        public override void SetLength(long value) => throw new NotSupportedException();
+
+        /// <summary>Releases the unmanaged resources used by the <see cref="ZstandardStream" /> and optionally releases the managed resources.</summary>
+        /// <param name="disposing"><see langword="true" /> to release both managed and unmanaged resources; <see langword="false" /> to release only unmanaged resources.</param>
+        protected override void Dispose(bool disposing)
+        {
+            try
+            {
+                if (disposing && _stream != null)
+                {
+                    if (_mode == CompressionMode.Compress)
+                    {
+                        WriteCore(ReadOnlySpan<byte>.Empty, isFinalBlock: true, checkActiveRWOps: false);
+                    }
+
+                    if (!_leaveOpen)
+                    {
+                        _stream.Dispose();
+                    }
+                }
+            }
+            finally
+            {
+                ReleaseStateForDispose();
+                base.Dispose(disposing);
+            }
+        }
+
+        /// <summary>Asynchronously releases the unmanaged resources used by the <see cref="ZstandardStream" />.</summary>
+        /// <returns>A task that represents the asynchronous dispose operation.</returns>
+        public override async ValueTask DisposeAsync()
+        {
+            try
+            {
+                if (_stream != null)
+                {
+                    if (_mode == CompressionMode.Compress)
+                    {
+                        await WriteCoreAsync(ReadOnlyMemory<byte>.Empty, CancellationToken.None, isFinalBlock: true).ConfigureAwait(false);
+                    }
+
+                    if (!_leaveOpen)
+                    {
+                        await _stream.DisposeAsync().ConfigureAwait(false);
+                    }
+                }
+            }
+            finally
+            {
+                ReleaseStateForDispose();
+                await base.DisposeAsync().ConfigureAwait(false);
+            }
+        }
+
+        private void ReleaseStateForDispose()
+        {
+            _stream = null!;
+
+            if (_encoderOwned)
+            {
+                _encoder?.Dispose();
+                _decoder?.Dispose();
+            }
+            else
+            {
+                _encoder?.Reset();
+                _decoder?.Reset();
+            }
+
+            byte[] buffer = _buffer;
+            if (buffer != null)
+            {
+                _buffer = null!;
+                if (!_activeRwOperation)
+                {
+                    ArrayPool<byte>.Shared.Return(buffer);
```

Comment:
There's synchronization around `_activeRwOperation` being updated concurrently.
Should we be doing something similar to guard against the race with dispose and other ops that might still be using the buffer?